File: src/Calculator.fs
namespace StringCalculator

open System

type Calculator() =

    static member private ParseDelimitersAndNumbers(input: string) =
        if input.StartsWith("//") then
            let idx = input.IndexOf('\n')
            let delimiterPart = input.Substring(2, idx - 2)
            let numbersPart = input.Substring(idx + 1)

            let customDelims =
                if delimiterPart.Contains("[") then
                    delimiterPart.Split([| '['; ']' |], StringSplitOptions.RemoveEmptyEntries)
                else
                    [| delimiterPart |]

            (customDelims, numbersPart)
        else
            ([| ","; "\n" |], input)

    static member private ParseNumbers(input: string, delimiters: string[]) =
        input.Split(delimiters, StringSplitOptions.RemoveEmptyEntries)
        |> Array.map int

    static member private ValidateNegatives(numbers: int[]) =
        let negatives = numbers |> Array.filter (fun n -> n < 0)
        if negatives.Length > 0 then
            let message = negatives |> Array.map string |> String.concat ", "
            raise (Exception($"negatives not allowed: {message}"))

    static member Add(numbers: string) : int =
        match numbers with
        | null | "" -> 0
        | _ ->
            let delimiters, numbersPart = Calculator.ParseDelimitersAndNumbers(numbers)
            let parsedNumbers = Calculator.ParseNumbers(numbersPart, delimiters)
            Calculator.ValidateNegatives(parsedNumbers)
            parsedNumbers
            |> Array.filter (fun n -> n <= 1000)
            |> Array.sum

File: python/src/calculator.py
class Calculator:
    def __init__(self, numbers: str):
        self.numbers = numbers or ""

    def _parse_delimiters_and_numbers(self):
        if self.numbers.startswith("//"):
            parts = self.numbers.split("\n", 1)
            delimiter = parts[0][2:]
            num_part = parts[1]
        else:
            delimiter = ","
            num_part = self.numbers.replace("\n", ",")
        return delimiter, num_part

    def _parse_numbers(self, input_str: str, delimiter: str):
        number_strings = input_str.split(delimiter)
        numbers = []
        for n in number_strings:
            if n:
                numbers.append(int(n))
        return numbers

    def add(self) -> int:
        if not self.numbers:
            return 0

        delimiter, num_part = self._parse_delimiters_and_numbers()
        parsed_numbers = self._parse_numbers(num_part, delimiter)

        return sum(parsed_numbers)
